### PR TITLE
feat(tui): add spinner animations for transitional download states

### PR DESCRIPTION
**Pull request by huseynovvusal**
## Description
This PR improves the perceived responsiveness of the Surge TUI by replacing static text indicators (`⋯`, `⏸`, `▶`) with smooth, continuously animating spinners during short-lived transitional states ([Queued](cci:2://file:///home/huseynvovvusal/Documents/Open%20Source/Surge/internal/engine/events/events.go:120:0-126:1), `Pausing...`, `Resuming...`). 

Previously, users might have misconstrued these statically-rendered states as the application hanging or freezing during asynchronous connection handshakes. This introduces a unified animated `spinner.Model` to provide clear, active feedback.

## How Has This Been Tested?
- Ran the internal TUI test suite via `go test ./internal/tui/...` which passes cleanly without structural regressions.
- Booted up `go run main.go` to test rendering behavior:
  - Queued multiple large downloads simultaneously to verify the active `⠋ Queued` spinner renders correctly while workers are busy or establishing connection probes.
  - Paused an active download to test the `⠋ Pausing...` fluid animation.
  - Resumed an active download to test the `⠋ Resuming...` fluid animation.
- Verified frame dimensions do not overflow existing margin boundaries.

## Screenshots
image

_Closes #280_

<h3>Greptile Summary</h3>

This PR replaces static Unicode glyphs (`⋯`, `⏸`, `▶`) with a shared animated `spinner.Model` for the `Queued`, `Pausing...`, and `Resuming...` transitional states in the Surge TUI, improving perceived responsiveness during async handshakes. The implementation is architecturally sound — a single root-level spinner is shared across all download items rather than one per download, and a cached `queuedSpinnerStyle` avoids per-tick allocations for the queued state.

Key changes:
- `components/status.go`: new `RenderWithSpinner` method and a cache-backed `queuedSpinnerStyle` built inside `rebuildStatusCache`
- `model.go`: adds `spinner.Model` to `RootModel`, initialised once and ticked from `Init()`
- `update.go`: new `spinner.TickMsg` handler with a `needsSpinner` guard to skip `UpdateListItems` when no animation is active — however, the next-tick command is **always** scheduled, so the ticker runs indefinitely even when all downloads are stable
- `list.go` / `view.go`: `spinnerView` string is threaded into item rendering at `UpdateListItems` time and passed directly to `renderFocusedDetails` on every `View()` call
- Tests added for `RenderWithSpinner`, `getDownloadStatus`, and `DownloadItem.Description` — but only for the three transitional states; completed/paused/error paths are not covered, and `list_test.go` declares a duplicate ANSI-stripping regex already present in the same package

<h3>Confidence Score: 3/5</h3>

- Safe to merge with minor fixes — the spinner loop never terminates and test coverage has gaps, but there are no data-loss or crash-level risks.
- The core spinner integration is correct and well-structured. The main functional concern is that the spinner tick command is unconditionally re-scheduled on every tick regardless of the `needsSpinner` guard, causing perpetual CPU wakeups even when no animation is needed. Test coverage for terminal states (done, paused, error) is absent, and a duplicate regex definition exists in the test files. None of these are blocking correctness issues, but the infinite ticker is a real inefficiency that the PR's own guard was intended to prevent.
- `internal/tui/update.go` — spinner tick command is always re-scheduled; `internal/tui/list_test.go` and `internal/tui/view_test.go` — incomplete edge-case coverage and duplicate regex

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| internal/tui/components/status.go | Adds `RenderWithSpinner` method and a cached `queuedSpinnerStyle`; the cache rebuild hooks are correctly wired and the read-lock pattern matches the existing mutex usage. |
| internal/tui/components/status_test.go | New `TestStatusRenderWithSpinner` test covers the queued spinner path and the non-queued pass-through; only happy paths are exercised. |
| internal/tui/list.go | Adds `spinnerView` field to `DownloadItem` and threads the current spinner frame into `UpdateListItems`; pausing/resuming branches still allocate fresh lipgloss styles per tick. |
| internal/tui/list_test.go | New `TestDownloadItem_Description` tests transitional states only; missing coverage for done/paused/error states, and declares a duplicate `testAnsiEscapeRE` already present as `ansiEscapeRE` in the same package. |
| internal/tui/model.go | Adds a single shared `spinner.Model` to `RootModel` with a clear justifying comment; `Init()` correctly issues the first `spinner.Tick` command. |
| internal/tui/update.go | Handles `spinner.TickMsg` with a `needsSpinner` guard for `UpdateListItems`, but the next-tick command is unconditionally scheduled regardless of `needsSpinner`, so the spinner loop never terminates. |
| internal/tui/view.go | Passes `spinnerView` into `renderFocusedDetails` and `getDownloadStatus`; the detail pane always uses a fresh `m.spinner.View()` on every render, keeping it in sync. |
| internal/tui/view_test.go | New `TestGetDownloadStatus` covers the three transitional states but leaves done, paused, and error paths untested, which are key non-spinner fallback routes. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Init as RootModel.Init()
    participant Tea as BubbleTea Runtime
    participant Update as Update()
    participant Spinner as spinner.Model
    participant List as UpdateListItems()
    participant View as View() / renderFocusedDetails()

    Init->>Tea: spinner.Tick (first tick cmd)
    Tea->>Update: spinner.TickMsg
    Update->>Spinner: spinner.Update(msg) → nextCmd, updated frame
    Update->>Update: needsSpinner? (scan downloads for pausing/resuming/queued)
    alt needsSpinner = true
        Update->>List: UpdateListItems() — snapshot spinner.View() into each DownloadItem
        Update->>Tea: Batch(nextCmd, ...)
    else needsSpinner = false
        Note over Update,Tea: nextCmd still appended — ticker runs forever
        Update->>Tea: Batch(nextCmd, ...)
    end
    Tea->>Update: spinner.TickMsg (next frame)
    View->>Spinner: m.spinner.View() — fresh frame for detail pane on every render
    View->>View: renderFocusedDetails(selected, w, spinnerView)
```

<details open><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `internal/tui/list.go`, line 10-21 ([link](https://github.com/surge-downloader/surge/blob/586da0541338908da475b55e772dc272cccdf8b3/internal/tui/list.go#L10-L21)) 

   <a href="#">P2</a> **Benchmark no longer exercises the new spinner code path**

   `BenchmarkDownloadDelegateRender` in `list_test.go` constructs a `DownloadItem` with `spinnerView: ""` (zero value) and a download with `Speed > 0`, so it always exercises `StatusDownloading → RenderWithSpinner → Render()` — completely bypassing the new `pausing`, `resuming`, and `StatusQueued` branches introduced in this PR.

   Adding a benchmark variant for the queued/pausing state would surface any per-tick allocation regressions before they reach production:

   ```go
   dQueued := DownloadItem{
       download:    &DownloadModel{ID: "q1", Filename: "ubuntu.iso", Total: 1 << 30},
       spinnerView: "⠙",
   }
   ```

   **Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
   
   <details><summary>Prompt To Fix With AI</summary>
   
   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/list.go
   Line: 10-21
   
   Comment:
   **Benchmark no longer exercises the new spinner code path**
   
   `BenchmarkDownloadDelegateRender` in `list_test.go` constructs a `DownloadItem` with `spinnerView: ""` (zero value) and a download with `Speed > 0`, so it always exercises `StatusDownloading → RenderWithSpinner → Render()` — completely bypassing the new `pausing`, `resuming`, and `StatusQueued` branches introduced in this PR.
   
   Adding a benchmark variant for the queued/pausing state would surface any per-tick allocation regressions before they reach production:
   
   ```go
   dQueued := DownloadItem{
       download:    &DownloadModel{ID: "q1", Filename: "ubuntu.iso", Total: 1 << 30},
       spinnerView: "⠙",
   }
   ```
   
   **Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))
   
   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: internal/tui/list_test.go
Line: 12

Comment:
**Duplicate ANSI regex in same package**

`testAnsiEscapeRE` in `list_test.go` compiles an identical pattern to `ansiEscapeRE` already defined in `view_test.go` (line 14). Because both files are in `package tui`, `list_test.go` can reference the existing symbol directly — no need for a second declaration.

```suggestion
// testAnsiEscapeRE removed — use ansiEscapeRE declared in view_test.go
```

Then replace the four usages of `testAnsiEscapeRE` in this file with `ansiEscapeRE`.

**Rule Used:** What: Eliminate duplicate logic, functions, or cod... ([source](https://app.greptile.com/review/custom-context?memory=f0a796a9-684f-4dfb-a5cf-8c99c16b63a1))

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/tui/view_test.go
Line: 45-93

Comment:
**Missing edge-case test coverage for terminal states**

`TestGetDownloadStatus` only exercises the three transitional states (`pausing`, `resuming`, `queued`). The `done`, `paused`, and `error` states go untested, leaving gaps where a regression in `RenderWithSpinner` (which falls back to `s.Render()` for non-queued statuses) or `getDownloadStatus` could go undetected.

Consider adding cases such as:

```go
{
    name: "Completed State",
    model: &DownloadModel{done: true},
    expected: "Completed",
},
{
    name: "Paused State",
    model: &DownloadModel{paused: true},
    expected: "Paused",
},
{
    name: "Error State",
    model: &DownloadModel{err: errors.New("timeout")},
    expected: "Error",
},
```

These ensure the non-spinner path through `getDownloadStatus` is also regression-protected.

**Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/tui/list_test.go
Line: 14-62

Comment:
**Missing edge-case test coverage for terminal states**

`TestDownloadItem_Description` mirrors the gap in `view_test.go` — only the three transitional states are exercised. The `done`, `paused`, and `err != nil` paths through `Description()` are unverified, meaning regressions in the `else` branch (which calls `status.RenderWithSpinner`) for completed/paused/error downloads won't be caught.

Add cases for at minimum `done: true`, `paused: true`, and a non-nil `err` to align with the other edge-case requirements for this PR.

**Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/tui/update.go
Line: 360-375

Comment:
**Spinner tick cmd always scheduled regardless of `needsSpinner`**

`m.spinner.Update(msg)` returns the next tick command in `cmd`, and it is unconditionally appended to `cmds` before the `needsSpinner` check runs. As a result the spinner loop never stops — it fires every tick interval forever, even when every download is in a stable completed/downloading/paused state and no animation is needed. The `needsSpinner` guard only prevents `UpdateListItems()` from running, not the next tick from being scheduled.

To actually stop the ticker when no animation is required, only append `cmd` when it is needed:

```go
case spinner.TickMsg:
    var cmd tea.Cmd
    m.spinner, cmd = m.spinner.Update(msg)

    needsSpinner := false
    for _, d := range m.downloads {
        if d.pausing || d.resuming || components.DetermineStatus(d.done, d.paused, d.err != nil, d.Speed, d.Downloaded) == components.StatusQueued {
            needsSpinner = true
            break
        }
    }
    if needsSpinner {
        cmds = append(cmds, cmd)
        m.UpdateListItems()
    }
    return m, tea.Batch(cmds...)
```

You would then need to restart the ticker (e.g. `m.spinner.Tick`) whenever a download transitions into a spinner-requiring state.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Last reviewed commit: ["test: add unit tests..."](https://github.com/surge-downloader/surge/commit/40d81fe535be5ef4ed9b83d2a3fec7195c3c7f00)</sub>

> Greptile also left **2 inline comments** on this PR.

**Context used:**

- Rule used - What: Eliminate duplicate logic, functions, or cod... ([source](https://app.greptile.com/review/custom-context?memory=f0a796a9-684f-4dfb-a5cf-8c99c16b63a1))
- Rule used - What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

### CODE REVIEW

**Comment by SuperCoolPencil** [comment]
While we can appreciate AI-assisted PRs, PRs where it seems the author has no idea what the code does will be rejected.

Leaving `[Insert Issue Number If Applicable]`  and the boilerplate instructions for screenshots in the description is a clear giveaway that this was copy-pasted without review.

**Comment by huseynovvusal** [comment]
> While we can appreciate AI-assisted PRs, PRs where it seems the author has no idea what the code does will be rejected.
> 
> Leaving `[Insert Issue Number If Applicable]`  and the boilerplate instructions for screenshots in the description is a clear giveaway that this was copy-pasted without review.

I have understanding of the codebase and 100% responsible of what I have implemented. I just wrote the description using AI quickly. You didn't even have a PR template, so not to keep it empty just added description. 

I feel bad, because I spent almost 3 hours on it and being judged due to a dumb PR description is not fair.

Additionally, I knew there was such issues in the PR, and planned to update/fill it/them after resolving AI reviewer feedbacks. You could have asked whether the PR is ready to merge or not, before being disrespectful to me.

**Comment by SuperCoolPencil** [comment]
I'm sorry, I genuinely didn't mean to be rude. I just looked at the PR description with the leftover prompt text, and combined with the code issues the bot flagged, it immediately felt like a zero-effort AI-generated PR. I jumped to conclusions.

I appreciate the time you put into the PR, and it makes complete sense that you used a generator for the description since we don't have a PR template set up yet. That's my fault for not having one in the repo.

Take your time with the bot's feedback. Just ping me here, and I will be happy to give it a proper human review.

**Comment by huseynovvusal** [comment]
> I'm sorry, I genuinely didn't mean to be rude. I just looked at the PR description with the leftover prompt text, and combined with the code issues the bot flagged, it immediately felt like a zero-effort AI-generated PR. I jumped to conclusions.
> 
> I appreciate the time you put into the PR, and it makes complete sense that you used a generator for the description since we don't have a PR template set up yet. That's my fault for not having one in the repo.
> 
> Take your time with the bot's feedback. Just ping me here, and I will be happy to give it a proper human review.

Thanks for your consideration. Of course, I was waiting to complate all the Greptile Reviews in order to add screenshoots and make some fixtures to PR. As you can see PR is not even synced with the codebase. It is also my fault sorry, because didn't wrote anywere that it is not final version. Appreciate your comment!

**Comment by huseynovvusal** [comment]
@SuperCoolPencil, I resolved AI reviewer comments and refactored the PR description. Could you please check?

**Comment by huseynovvusal** [comment]
@SuperCoolPencil, here they are! I have implemented the changes you wanted.

**Review by SuperCoolPencil — APPROVED** [review]
Awesome! 

The tests can be tightened up, but this is good for now

**Comment by SuperCoolPencil** [comment]
oh um, You requested review so I assumed you were done...

**Comment by huseynovvusal** [comment]
> oh um, You requested review so I assumed you were done...

Thanks for merging. Before you merge I added more tests. I requested review almost 1 hr ago I guess. So, never mind

**Comment by SuperCoolPencil** [comment]
Well you addressed my comment anyway, so everything's good now

File: internal/tui/update.go [hunk 1]
```diff
@@ -355,6 +357,23 @@ func (m RootModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
...
+		for _, d := range m.downloads {
+			if d.pausing || d.resuming || components.DetermineStatus(d.done, d.paused, d.err != nil, d.Speed, d.Downloaded) == components.StatusQueued {
+				needsSpinner = true
+				break
+			}
+		}
+		if needsSpinner {
+			m.UpdateListItems()
+		}
+		return m, tea.Batch(cmds...)
```
[the remark this answers is not in the document]
Comment (SuperCoolPencil) [reply internal/tui/update.go:375]: Well, this needs to be addressed @huseynovvusal

[the remark this answers is not in the document]
Comment (huseynovvusal) [reply internal/tui/update.go:375]: Ok, on the same branch I will resolve it. Thanks for mentioning.
